sympy/ntheory/qs.py

```diff
@@ -0,0 +1,412 @@
+    matrix = []
+    for s_relation in smooth_relations:
+        matrix.append(s_relation[2])
+    return matrix
+
+def _gauss_mod_2(A):
+    """Fast gaussian reduction for modulo 2 matrix.
+
+    Parameters:
+    ===========
+
+    A : Matrix"""
+    import copy
+    matrix = copy.deepcopy(A)
+    row = len(matrix)
+    col = len(matrix[0])
+    mark = [False]*row
+    for c in range(col):
+        for r in range(row):
+            if matrix[r][c] == 1:
+                break
+        mark[r] = True
+        for c1 in range(col):
+            if c1 == c:
+                continue
+            if matrix[r][c1] == 1:
+                for r2 in range(row):
+                    matrix[r2][c1] = (matrix[r2][c1] + matrix[r2][c]) % 2
+    dependent_row = []
+    for idx, val in enumerate(mark):
+        if val == False:
+            dependent_row.append([matrix[idx], idx])
+    return dependent_row, mark, matrix
+
+def _find_factor(dependent_rows, mark, gauss_matrix, index, smooth_relations, N):
+    idx_in_smooth = dependent_rows[index][1]
+    com_u = [smooth_relations[idx_in_smooth][0]]
+    com_v = [smooth_relations[idx_in_smooth][1]]
+    dept_row = dependent_rows[index][0]
+
+    for idx, val in enumerate(dept_row):
+        if val == 1:
+            for row in range(len(gauss_matrix)):
+                if gauss_matrix[row][idx] == 1 and mark[row] == True:
+                    com_u.append(smooth_relations[row][0])
+                    com_v.append(smooth_relations[row][1])
+                    break
+
+    u = 1
+    v = 1
+    for i in com_u:
+        u *= i
+    for i in com_v:
+        v *= i
+    #assert u**2 % N == v % N
+    v = sqrt_int(v)
+    return igcd(u - v, N)
+
+def sqrt_int(n):
+    """Return the square root of the given integer, rounded down to the
+    nearest integer.
+    """
+    from math import floor, log2
+    a = n
+    s = 0
+    o = 1 << (floor(log2(n)) & ~1)
+    while o != 0:
+        t = s + o
+        if a >= t:
+            a -= t
+            s = (s >> 1) + o
+        else:
+            s >>= 1
+        o >>= 2
+    assert s*s == n
+    return s
+
+def qs(N, B, M, ERROR_TERM=25*2**10, threshold=10, seed=1234):
+    random.seed(seed)
+    factor_base = _generate_factor_base(B, N)
+    smooth_relations = []
+    poly_count = 0
+    partial = {}
+    proper_factor = set()
+    while True:
+        if poly_count == 0:
+            sieve_poly_i, B_array = _initialize_first_polynomial(N, M, factor_base)
+        else:
+            sieve_poly_i = _initialize_ith_poly(N, factor_base, poly_count, sieve_poly_i, B_array)
+        poly_count += 1
+        if poly_count >= 2**(len(B_array) - 1):
+            poly_count = 0
+        sieve_array = _gen_sieve_array(M, factor_base)
+        s_rel, p_f = _trial_division_stage(N, M, factor_base, sieve_array, sieve_poly_i, partial, ERROR_TERM)
+        smooth_relations += s_rel
+        proper_factor |= p_f
+        if len(smooth_relations) >= len(factor_base) + threshold:
```

Comment:
If the `num_primes`         and `M` are too small then this will loop forever, so it might be possible to make a function to calculate `num_primes` and `M` instead of making this as parameter as eventually `qs` will be used in `factor_int`. But we can also leave these as parameters for users to select cause the estimation of these parameters is known till around 70-80 digits numbers.

Reply:
> If the num_primes and M are too small then this will loop forever

Then this needs to be taken care of.

Reply:
I am thinking of writing a function through which then factor_base size and M can be selected.